lib/streamlit/runtime/state/widgets.py

```diff
@@ -243,20 +244,23 @@ def coalesce_widget_states(
         wstate.id: wstate for wstate in new_states.widgets
     }
 
-    trigger_value_types = [("trigger_value", False), ("string_trigger_value", None)]
+    trigger_value_types = [
+        ("trigger_value", False),
+        ("string_trigger_value", StringTriggerValueProto(data=None)),
+    ]
     for old_state in old_states.widgets:
         for trigger_value_type, unset_value in trigger_value_types:
             if (
                 old_state.WhichOneof("value") == trigger_value_type
-                and old_state.trigger_value != unset_value
+                and getattr(old_state, trigger_value_type) != unset_value
             ):
                 # Ensure the corresponding new_state is also a trigger;
                 # otherwise, a widget that was previously a button but no longer is
                 # could get a bad value.
                 new_trigger_val = states_by_id.get(old_state.id)
                 if (
-                    new_trigger_val
-                    and new_trigger_val.WhichOneof("value") == trigger_value_type
+                    not new_trigger_val
+                    or new_trigger_val.WhichOneof("value") == trigger_value_type
```

Comment:
The bug here ends up being extremely tricky to test via unit test, so I just ended up adding an e2e test for this.

The issue is that the `WidgetStateManager` on the frontend immediately deletes trigger values from the widget state it's keeping track of after it calls its `onWidgetValueChanged` callback (which sends a `BackMsg` from client to server). Because of this, we never actually run into this code in practice since it will never be the case that `old_state` contains a set trigger value for some widget ID and `new_state` contains an unset trigger value for the same widget ID. Unless the user interacted with the same widget a second time, the trigger value will always have been cleared from `new_state`.

Reply:
Great investigation. I think it might be valuable to leave that^ (in a shortened form) as a comment in the code.
For my understanding: did we not run into any issues here (since it never happened), because so far we always had full app runs were this scenario did not exist anyways?

Reply:
Possibly. I do think that it could sometimes be surprising that button clicks get lost in full app reruns too, but it's a lot harder to hit these scenarios compared to when running fragments, and at least the behavior is justifiable and (as far as I know) we haven't had many people (if any) complain about it. I'll add an explanatory comment for this.